### DISCUSSION TITLE
Seeking Newbie Level Assistance With Training

### DISCUSSION THREAD

**Discussion by GeorgiaM-honestly** [oobabooga/text-generation-webui#5243] [General]
Hello!

I am seeking newbie level assistance with training.

Use Case: Some technical knowledge that could probably be saved as a raw text file

Desired Result: Be able to use normal language to ask for exact (rather than creative) information about information in this raw text file, which would be used to train a LORA

What I tried: Loaded mistralai_Mistral-7B-v0.1, then created a raw text file with the information (just over 1,000 lines to start with) in the training/datasets directory. In the training tab I started with the default settings, pointed the raw text tab at the file, and trained.

After this I re-loaded the model, refreshed the loras, and selected the new lora and loaded it.

I then went to the chat tab, set it to chat-instruct which the model loading tab had said I should do (either that or "instruct", then asked it about some specific subjects that were in the raw input text file.

While sometimes I get hints of this data, for the most part, there is nothing related. I then developed a, for lack of a better word because I don't know what this is called, an intro query / information for the given chat, giving it clues about what i was looking for, that I was looking for exact outputs rather than creative outputs, and so on.

This increased how often I got expected results, but not very much.

At this point I feel pretty frustrated, and I feel that I don't know what I'm doing yet. I have looked at the documentation on the wiki for the training tab and that does make sense to me, yet I feel I am still missing something.

Could people point me at some more / other documentation that might help? I have found some of my own however it seems that this tech moves SO fast that much of it is out of date if not wrong (not by intention) by the time I find it.

Thank you for reading!

**Comment by araleza** [comment]
Hi there,

I'm no great expert, but I think I'm one of the few people doing LoRA training via ooba webui, so I decided I'd share my thoughts here.  Some of the stuff I'm going to write here won't be specifically relevant to you, but I imagine a lot of people want to try LoRA training for various other reasons.  (TLDR: Use RAG instead of LoRA for adding knowledge to a model).  My hardware is a 24 GB Nvidia 4090 RTX.

(@oobabooga - is any of what I've written here any use as docs?)

There seems to be various things people try to do with LoRAs:

1) Training for style/personality
2) Training to add factual information
3) Training on a massive new dataset, to generally improve the model's flexibility, or add roleplay abilities
4) Training to remove alignment (i.e. refusals to answer various questions)

I've attempted both (1) and (2) from these, using raw files in each case, so that's what I'll be describing here.

The reason I'm doing LoRA training is both so I can more precisely control the personality and sound of the person I'm trying to produce, and to save valuable 'context' for the LLM.  I only use the context to provide facts to the character, with the personality coming from the LoRA training.  Previously, I did attempt to provide facts via the LoRA training, and I'll get to that shortly.

In order to load a model for LoRA training, it has to be loaded with the Transformers model loader, not any of the other ones.  You can apply the LoRA to the other ones after you've done training with the Transformers model loader though.  You can even load 4-bit GPTQ files with the Transformers model loader if you tick 'auto-devices' and 'disable_exllama'.  Currently, there's a bug that stops you applying LoRAs with models loaded with the Transformers model loader, but training will still work:

   https://github.com/oobabooga/text-generation-webui/issues/5182

My raw file just contains replies for the person whose personality I'm attempting to be learned, separated by '---' lines, because for some reason the default '\n\n\n' doesn't seem to work right.  I think there's a bug there.  After starting a LoRA training session, you can open a file called 'logs/train_dataset_sample.json' inside the text-generation-webui directory, and that will show you examples from your data of what's actually being given to the model to train with.

Achieving the first goal seems fairly simple.  When you run the LoRA training, the loss begins to drop.  At some point, the model you're training starts to 'sound like' the raw text file.  You don't want to push the loss too low, or it'll only say stuff from the file though, and then it'll start to break entirely.  Watch out for the default number of epochs being set to 3.  This number is set up for people who are trying to train on massive datasets.  If you have a smaller amount of training data, your epoch count should be significantly higher.

Also, I found I had to have a much lower training rate than the Training tab suggested (I sometimes use 1e-5 for drafting LoRAs, and tend to use 2e-6 or lower, right from the start of training, to produce a finished version).  Having the learning rate too high tends to cause looping or repetitive replies when using that LoRA.  Although LLMs have a habit of repeating stuff anyway.

But you're not training for style/personality, you're trying to add facts.

I tried adding facts with the technique I described above, and it was close to being a total failure.  I tried to get my character to learn how old I was by writing example text that referenced my age, but when I asked the character my age in chat, it made up random ages.  Same for training for the city I live in.  The character would imagine London, or New York, even though I do not live in either of those places.  It would sometimes get my city right (more often than it got my age right), so it was picking up something, but it was very weak.  Similar to your experience, I think.

Perhaps people have more luck in adding facts when they follow the Vicuna / Alpaca form of Question/Answer pairs?  If you don't know what those are, you should look them up.  And you should also look up which one your model is using.  Ooba webui sometimes states which one is being used when you load the model.  Basically these are fixed forms of text that include phrases like '### Instruction', or '### Response' etc., which the models were originally trained with.  If you transform your own raw file into this format, the model may learn much more effectively, as the pattern of text now matches its existing data.

Rather than just reformatting your raw files into this form, which will involve repeating '### Instruction' etc. many times, you can instead reformat it into a .json file.  The LoRA training tab has a 'Formatted Dataset' tab inside it while takes .json instead.  That basically allows it to pick up json tags like 'instruction', 'input', and 'output', and converts these to '### Instruction', '### Input', '### Response' on the fly if that's what the particular model you're training wants.  You can see what's converted into what by looking at files like 'training/format/alpaca-format.json'.

But even still, I hear that adding knowledge with a LoRA can be tricky.  There are other options.  Firstly, some people are using something called RAG instead.  RAG allows the model to see your question, and then look up another information source (e.g. a web page you've written with answers to the questions on it), and then it inserts that into its own context as if you'd just typed all that information in before you asked your question.  Text in the context is far more effective at adding knowledge, and it doesn't require any modification to the model.  Of course you have to look at the maximum context size your model can handle.  If it's 4096 tokens, you need to break the web pages up into small enough sections that you don't use up all the context before the model can reply.  Which means you need even more RAG code to help it pick the right web page with the answer to include.

If you really want to push hard down the route of modifying the model to have additional knowledge, there are research papers that have investigated various techniques for actually adding facts to an existing model, or even deleting facts from a model.  They're hard work though.  You can find lots of Reddit threads with lots of links to follow if you're determined to try this, here:

   https://old.reddit.com/r/LocalLLaMA/search?q=knowledge&restrict_sr=on

But for what you're trying to do, I'd recommend going with RAG instead.

**Reply by GeorgiaM-honestly** [reply]
@araleza , THANK YOU so much for the lengthy and detailed response! This is very helpful. I believe I'm starting to come to terms with, and let me know what you (or anyone else) think: That I'm probably looking at creating my own model. That seems daunting given everything I've read about making LLMs, however, I don't think the first L (Large) applies in my case. It's not books worth of info much less a library of books. Which seems to be a good thing as those are INCREDIBLY RESOURCE HUNGRY tasks. 

I took some notes about 3 years ago when I did something similar, however, I can't seem to find them. Perhaps more irritatingly, general web searches about how to do it turn up a myriad of things that don't seem to be what I'm looking for. So at this point I believe part of my issue is I don't know the terms to search for it.

This is clearly veering away from the subject of the post, but what do you think?

**Comment by FartyPants** [comment]
You can't write one Q/A on a "fact" and expect the model will suddenly be able to use that. There is million of instances in pre-training where the answer in context was different than yours. Your Q/A does not magically override that. That's not how LLM works, nor what finetuning is for. 

Finetuning is to give a form so the model knows what game you are playing. So let's say 100 varying question where the answer is in a HTML table teaches the model that for things that look like your questions it should respond in a html table. The model of course already knows what html table is - hence it is much harder to teach it in terms of never seen things - like "blimp-blup" table - which I just made up. You'd need much more examples of such thing and the model will likely still made much more mistakes because it never ever saw blimp-blup table.

Another way you can look at this problem - try to teach a Model made up language. You will not succeed well in finetuning, because even if you think you have very precise rules for your made up language the model will still be making stuff up, breaking your rules. 

Another issue with this is - teach your model something false - WW2 ended in 1955 and the effect will be seen in many other responses - because you just taught the model that you want it to lie. That the answers you desire are NOT the ones it should give you. Hence: Lora that is contrary to common wisdom (even creative such as "we have bases on the moon") can as well hurt the model somewhere else.
 
If you really want to hammer the age of a character down (or concrete info), you would need to write many, many Q/A that talk about age of that character in different context. (but note the above, it will hurt the model somewhere else) The model will learn that if you talk about age and that name of character then usual answer is XX. It may also figure out that other people age is XX as well or close. 
Lora is extremely unsuited for this type of learning. You can hammer knowledge, but it is painful and literally the wrong place to do so. (Picture Fry from Futurama hammering nail with a nail)

In fact if you write an example Q/A (or text) as your pre-prompt where you specify age, most RP friendly models will actually use that much more readily than Lora. This is how RP like SillyTavern use world info and memories and authors note - in many RP models you often need to mention these fact only once at the beginning of chat - in a format the RP model is used to get facts (was finetuned in) and it works 100% better than finetuning. The bonus: you didn't destroy the model telling it contrary info.
